lib/core/util.c

```diff
@@ -422,10 +422,26 @@ h2o_iovec_t h2o_build_destination(h2o_req_t *req, const char *prefix, size_t pre
     }
 
     /* append suffix path and query */
-    parts[num_parts++] = h2o_uri_escape(
-        &req->pool, req->path_normalized.base + req->pathconf->path.len, req->path_normalized.len - req->pathconf->path.len, "/@:");
-    if (req->query_at != SIZE_MAX)
-        parts[num_parts++] = h2o_iovec_init(req->path.base + req->query_at, req->path.len - req->query_at);
+
+    if (escape || req->res_is_delegated) {
+        /* When proxying, we don't want to modify the input URL, unless the url
+           has been rewritten already */
+        parts[num_parts++] = h2o_uri_escape(&req->pool, req->path_normalized.base + req->pathconf->path.len,
+                                            req->path_normalized.len - req->pathconf->path.len, "/@:");
+        if (req->query_at != SIZE_MAX) {
+            parts[num_parts++] = h2o_iovec_init(req->path.base + req->query_at, req->path.len - req->query_at);
+        }
+    } else {
+        if (req->input.path.len > 1) {
```

Comment:
How about using `req->path` instead of `req->input.path` and also removing the check of `req->res_is_delegated` in the if statement above?

That way, I assume we can use the unnormalized path for both before and after delegation (note: I believe the proxy handler should use the unnormalized path independent to how the URL is given (e.g. via an HTTP request or via a X-Reproxy-URL header)).